std/algorithm/iteration.d

```diff
@@ -4289,3 +4304,633 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Computes the $(WEB en.wikipedia.org/wiki/Binomial_coefficient, binomial coefficient)
+of n and k.
+It is also known as "n choose k" or more formally as $(D _n!/_k!(_n-_k)).
+
+Params:
+    n = non negative integer
+    k = non negative integer, k <= n
+
+Returns:
+    binomial coefficient
+*/
+
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
+in
+{
+    assert(n > 0, "binomial: n must be > 0.");
+}
+body
+{
+    if (k < 0 || k > n)
+        return 0;
+    if (k == n)
+        return 1;
+    if (k > (n / 2))
+        k = n - k;
+    size_t result = 1;
+    for (size_t i = 1, m = n; i <= k; i++, m--)
+        result = result * m / i;
+    return result;
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5, 2) == 10);
+    assert(binomial(6, 4) == 15);
+    assert(binomial(3, 1) == 3);
+}
+
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5, 1) == 5);
+    assert(binomial(5, 0) == 1);
+    assert(binomial(1, 2) == 0);
+    assert(binomial(1, 0) == 1);
+    assert(binomial(1, 1) == 1);
+    assert(binomial(2, 1) == 2);
+}
+
+/**
+Lazily computes the Cartesian power of $(D r) with itself
+for a number of repetitions $(D repeat).
+If the input is sorted, the product is in lexicographic order.
+For example $(D"AB".cartesianPower(2).array) returns $(D["AA", "AB", "BA", "BB"])
+
+Params:
+    r = $(REF_ALTTEXT Random access range, isRandomAccessRange, std, range, primitives)
+    repeat = number of repetitions
+
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat = 1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(repeat >= 1, "Invalid repeat");
+}
+body
+{
+    import std.array: array;
+    import std.range: Indexed;
+    static struct CartesianPower
+    {
+    private:
+        Range _r;
+        size_t[] _state, _stateBack;
+
+        size_t _repeat, _nrElements;
```

Comment:
`_repeat` seems to never be used after the constructor (as it's identical to `_state.length` which is used instead)

---

`_nrElements` is `_r.length` - it's probably better to cache this as a local variable in the constructor and `pop*` functions respectively instead of as a field, i.e. as:
```d
void popFront()
{
    immutable nrElements = _r.length;
    /* use nrElements */
}
```
This is more in line with how other range algorithms do it. Chances are that `_r.length` is simply a struct field already, and when it isn't, it's probably cheap. We tend to err on the side of less struct state than aggressive caching.